kernel: implement some system calls

tyrantlizard77 opened the pull request:
This PR shows system calls implemented for thread, alert, pipe, timer, stack, message queue, and mutex kernel objects. A few other miscellaneous kernel objects also converted to support Philosophers demo in user mode.

Finally, we demonstrate a driver subsystem by converting the Sensor driver APIs.

Commit:
kernel: implement some more system calls

These are needed to demonstrate the Philosophers demo with threads
running in user mode.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
sensors: implement system calls

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: convert mutex APIs to system calls

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: convert alert APIs to system calls

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: convert most thread APIs to system calls

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: convert pipe APIs to system calls

k_pipe_block_put() will be done in another patch, we need to design
handling for the k_mem_block object.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: convert timer APIs to system calls

k_timer_init() registers callbacks that run in supervisor mode and is
excluded.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: convert stack APIs to system calls

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented on kernel/thread.c:
```diff
@@ -290,9 +324,40 @@ k_tid_t k_thread_create(struct k_thread *new_thread,
+	_SYSCALL_MEMORY(stack, stack_size + guard_size, 1, ssf);
+
+	/* User threads may only create other user threads */
+	margs->arg9 |= K_USER;
```
should this be an assert instead of an implicit assignment?

andrewboie commented on kernel/thread.c:
```diff
@@ -290,9 +324,40 @@ k_tid_t k_thread_create(struct k_thread *new_thread,
+	 * any guard areas
+	 */
+	guard_size = (u32_t)K_THREAD_STACK_BUFFER(stack) - (u32_t)stack;
+	_SYSCALL_MEMORY(stack, stack_size + guard_size, 1, ssf);
```
stacks are supposed to be opaque so maybe better to add an API to get the full size instead of calculating here

andrewboie commented on kernel/thread.c:
```diff
@@ -290,9 +324,40 @@ k_tid_t k_thread_create(struct k_thread *new_thread,
+					    (k_thread_entry_t)entry,
+					    (void *)p1, (void *)margs->arg6,
+					    (void *)margs->arg7, margs->arg8,
+					    margs->arg9, margs->arg10);
```
need to verify priority.
also thinking about DoS scenarios...does it make sense to restrict thread priority to preemptible values if the caller is preemptible?

andrewboie commented:
I removed k_thread_create() from this PR, needs discussion on what policies we are going to need for it. Rest of the syscalls here are straightforward.

Commit:
kernel: convert k_msgq APIs to system calls

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
rebased, use of __syscall_inline removed

Commit:
kernel: remove some kernel objects from tracking

These are removed as the APIs that use them are not suitable for
exporting to userspace.

- Kernel workqueues run in supervisor mode, so it would not be
appropriate to allow user threads to submit work to them. A future
enhancement may extend or introduce parallel API where the workqueue
threads may run in user mode (or leave as an exercise to the user).

- Kernel slabs store private bookkeeping data inside the
user-accessible slab buffers themselves. Alternate APIs are planned
here for managing slabs of kernel objects, implemented within the
runtime library and not the kernel.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
updated to include a patch to remove kernel workqueue and slab objects from tracking, as these APIs in their current incarnation are not suitable for being turned into system calls.

this patch pretty much defines what kernel object system call APIs we will have.

what remains is converting the rest of the driver subsystem APIs, and introducing userspace-compatible alternatives for slabs, workqueues, and poll objects

agross-oss reviewed (approved):
by an large this looks fine.  I'd prefer to have some sort of commit message per commit, but that's a nit.
